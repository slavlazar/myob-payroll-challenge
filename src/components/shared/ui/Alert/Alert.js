import React from 'react';
import PropTypes from 'prop-types';

import { Alert as AlertStyle, Header, Body } from './Alert.style';

/**
 * Renders a styled alert box based on size and type
 * @param {object} Object of title, size, type and children of Alert
 */
const Alert = ({ children, title, size, type }) => (
  <AlertStyle size={size} type={type}>
    {title && (
      <Header className="alert-header" type={type}>
        {title}
      </Header>
    )}
    <Body type={type} title={title}>
      {children}
    </Body>
  </AlertStyle>
);

Alert.propTypes = {
  /** Message to display in the rendered alert component */
  children: PropTypes.node.isRequired,
  size: PropTypes.oneOf(['small', 'large', 'default']),
  type: PropTypes.oneOf(['danger', 'success', 'default']),
  title: PropTypes.string,
};

Alert.defaultProps = {
  size: 'default',
  type: 'default',
  title: null,
};

export default Alert;
