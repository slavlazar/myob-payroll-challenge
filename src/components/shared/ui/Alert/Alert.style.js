import styled from 'styled-components';
import { darken } from 'polished';

const shadeColourType = ({ shadeValue, type, theme }) => {
  switch (type) {
    case 'danger':
      return darken(shadeValue, theme.colour.alert.danger);
    case 'success':
      return darken(shadeValue, theme.colour.alert.success);
    default:
      return darken(shadeValue, theme.colour.alert.default);
  }
};

export const Header = styled.div`
  position: relative;
  padding: 0.75em 1em;
  line-height: ${props => props.theme.lineHeight.heading};
  display: flex;
  align-items: center;
  justify-content: space-between;
  font-weight: ${props => props.theme.font.weight.bold};
  color: #ffffff;
  border-radius: 0.4rem 0.4rem 0 0;
  background-color: ${props =>
    shadeColourType({ shadeValue: 0.5, type: props.type, theme: props.theme })};
`;

export const Body = styled.div`
  padding: 1.25em 1.5em;
  border-style: solid;
  border-color: ${props =>
    shadeColourType({ shadeValue: 0.4, type: props.type, theme: props.theme })};
  color: ${props =>
    shadeColourType({ shadeValue: 0.6, type: props.type, theme: props.theme })};
  border-width: ${props => (props.title ? 0 : '0 0 0 0.4rem')};
  border-radius: ${props => (props.title ? '0 0 .4rem .4rem' : '0.4rem')};
`;

export const Alert = styled.article`
  font-size: ${props => {
    switch (props.size) {
      case 'small':
        return props.theme.font.size.small;
      case 'large':
        return props.theme.font.size.large;
      default:
        return props.theme.font.size.medium;
    }
  }};
  background-color: ${props => {
    switch (props.type) {
      case 'danger':
        return props.theme.colour.alert.danger;
      case 'success':
        return props.theme.colour.alert.success;
      default:
        return props.theme.colour.alert.default;
    }
  }};
  border-radius: 0.4rem;
  margin-bottom: 2rem;
`;
