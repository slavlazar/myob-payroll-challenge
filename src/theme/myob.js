const fontFamilyMYOBRegular = 'NeutroMYOB-Regular';
const fontFamilyMYOBMedium = 'NeutroMYOB-Medium';
const fontFamilyMYOBStrong = 'NeutroMYOB-Demi';

export default {
  colour: {
    primary: '#8241AA',
    secondary: '#ed005f',
    text: '#515e6d',
    heading: '#2e3e4f',
    background: '#f3f4f5',
    footerBackground: '#1b2830',
    navBackground: '#ffffff',
    buttonPrimary: '#ffffff',
    border: '#efefef',
    buttonPrimaryBackground: '#00aa65',
    alert: {
      default: '#f5f5f5',
      success: '#00d1b2',
      danger: '#fff5f7',
    },
  },
  font: {
    size: {
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      large: '1.8rem',
      xlarge: '2.2rem',
    },
    variants: {
      body: {
        family: fontFamilyMYOBRegular,
        style: 'normal',
        weight: 400,
        src: [
          'url(https://assets.digital.myob.com/fonts/NeutroMYOB-Regular.woff2)',
          'url(https://assets.digital.myob.com/fonts/NeutroMYOB-Regular.woff)',
        ],
      },
      heading: {
        family: fontFamilyMYOBMedium,
        style: 'normal',
        weight: 600,
        src: [
          'url(https://assets.digital.myob.com/fonts/NeutroMYOB-Medium.woff2)',
          'url(https://assets.digital.myob.com/fonts/NeutroMYOB-Medium.woff)',
        ],
      },
      strong: {
        family: fontFamilyMYOBStrong,
        style: 'normal',
        weight: 700,
        src: [
          'url(https://assets.digital.myob.com/fonts/NeutroMYOB-Demi.woff2)',
          'url(https://assets.digital.myob.com/fonts/NeutroMYOB-Demi.woff)',
        ],
      },
    },
    family: {
      body: fontFamilyMYOBRegular,
      heading: fontFamilyMYOBMedium,
      strong: fontFamilyMYOBStrong,
      fallback:
        '-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif',
    },
    weight: {
      bold: 600,
      normal: 400,
    },
  },
  lineHeight: {
    text: 1.7,
    heading: 1.5,
  },
  media: {
    large: 992, // mobile-first
    desktop: 768, // mobile-first
    tablet: 576, // mobile-first
  },
};
