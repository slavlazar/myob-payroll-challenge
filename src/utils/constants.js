export const LOGIN_URL = '/login';
export const API_LOGIN_ENDPOINT = '';
