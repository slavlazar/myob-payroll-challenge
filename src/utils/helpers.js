/**
 * Helper method to get file extension of given file name
 * @param {string} filename Full filename
 * @return {string} Given filename, will return the extension
 */
export const getFileExtension = filename =>
  filename
    .split('.')
    .pop()
    .replace(/[\W_]+/g, '');

/**
 * Helper method to general initial values from a given schema
 * @param {array} fields Array of fields to generate initial values
 * @return {object} Object of generated initial values from schema
 */
export const generateInitialValuesFromSchema = (fields = []) =>
  fields.reduce((accum, key) => ({ ...accum, [key]: '' }), {});

/**
 * Helper method to determine if given object is empty
 * @param {object} obj Object to check if empty
 * @return {boolean} Given object, return true if empty, false otherwise
 */
export const isObjectEmpty = obj =>
  Object.keys(obj).length === 0 && obj.constructor === Object;

/**
 * Helper method to log error messages to console
 * Curried function that first takes name of group,
 * followed by console messages
 * @param {string} groupName Name of group to log
 * @returns {[function, void]} Either returns a curried function or logs
 * to console
 */
export const logError = groupName => (...consoleMsg) => {
  console.group(groupName);
  consoleMsg.forEach(msg => {
    console.error(msg);
  });
  console.groupEnd();
};

/**
 * Helper method to round number to nearest integer
 * @param {number} val Number to round
 * @return {number} Rounded number
 */
export const round = val => Math.round(val);

/**
 *
 * @param {number} val Number to format
 * @param {boolean} roundValue Whether to round the value before formatting
 * @return {string} Formatted currency value
 */
export const formatCurrency = (val, roundValue = true) => {
  const valToConvert = roundValue ? round(val) : val;

  return Number(valToConvert).toLocaleString('en-AU', {
    style: 'currency',
    currency: 'AUD',
    minimumFractionDigits: 2,
  });
};
