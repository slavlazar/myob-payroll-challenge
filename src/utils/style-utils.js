import React from 'react';
import { mount } from 'enzyme';
import { css, ThemeProvider } from 'styled-components';

import theme from '../theme';

const { media: sizes } = theme;

// Media query, usage: ${media.tablet`padding: 1.5rem`}
// min-width used for mobile-first design
export const media = Object.keys(sizes).reduce((acc, label) => {
  acc[label] = (...args) => css`
    @media (min-width: ${sizes[label] / 16}em) {
      ${css(...args)}
    }
  `;

  return acc;
}, {});

export const renderWithTheme = component => {
  return mount(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
};
