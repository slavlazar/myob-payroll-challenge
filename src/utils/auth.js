import axios from 'axios';

import { API_LOGIN_ENDPOINT } from './constants';

/**
 * Login method that calls upon our API endpoint to log the user in
 * @param {object} Object of email and password keys to log user in with
 * @async
 */
const login = async ({ email, password }) => {
  const { data: userInfo } = await axios.post(API_LOGIN_ENDPOINT, {
    email,
    password,
  });

  return userInfo;
};

export default {
  login,
};
