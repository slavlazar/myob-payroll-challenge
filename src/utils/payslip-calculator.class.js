import moment from 'moment';

import * as TAXABLE_INCOME from '../config/tax-rates.json';

import { logError as logErrorGroup, formatCurrency } from './helpers';

/** Class representing Payslip calculator. */
class PayslipCalculator {
  static logError = logErrorGroup('PayslipCalculator');

  /**
   * Create a payslip calculator.
   * @param {string} firstName - Employee first name.
   * @param {string} lastName - Employee last name.
   * @param {number} annualSalary - Employee annual salary.
   * @param {number} superRate - Employee super rate.
   */
  constructor(employeeDetails) {
    ({
      firstName: this._firstName,
      lastName: this._lastName,
      annualSalary: this._annualSalary,
      superRate: this._superRate,
    } = employeeDetails);

    this._grossIncome = 0;
    this._incomeTax = 0;
    this._netIncome = 0;
    this._superTotal = 0;
    this._payFrequency = 'Monthly';

    this._determinePayPeriod(employeeDetails.payPeriod);
  }

  /**
   * Get employee full name.
   * @return {string} Employee first name and last name.
   */
  get name() {
    return `${this._firstName} ${this._lastName}`;
  }

  /**
   * Get employee pay period.
   * @return {string} Employee pay period based on month and year.
   */
  get payPeriod() {
    return this._payPeriod;
  }

  /**
   * Get employee pay frequency.
   * @return {string} Employee pay frequency - Defaults to monthly for demo.
   */
  get payFrequency() {
    return this._payFrequency;
  }

  /**
   * Get employee annual salary.
   * @return {string} Employee salary formatted to AUD currency.
   */
  get annualSalary() {
    return formatCurrency(this._annualSalary);
  }

  /**
   * Get employee gross income.
   * @return {string} Employee gross income formatted to AUD currency.
   */
  get grossIncome() {
    return formatCurrency(this._grossIncome);
  }

  /**
   * Get employee income tax.
   * @return {string} Employee income tax formatted to AUD currency.
   */
  get incomeTax() {
    return formatCurrency(this._incomeTax);
  }

  /**
   * Get employee net income.
   * @return {string} Employee net income formatted to AUD currency.
   */
  get netIncome() {
    return formatCurrency(this._netIncome);
  }

  /**
   * Get employee super total.
   * @return {string} Employee super total formatted to AUD currency.
   */
  get superTotal() {
    const superCalc = this._superRate / 100;
    return formatCurrency(this._grossIncome * superCalc);
  }

  /**
   * Get employee pay.
   * @return {string} Employee pay for the month, formatted to AUD currency.
   */
  get pay() {
    return formatCurrency(this._netIncome - this._superTotal);
  }

  /**
   * Get total tax amount for employee, given taxable amount,
   * tax rate and base tax amount.
   * @static
   * @param {number} Taxable amount of employee
   * @returns [{function}, {function}, {number}] Total tax amount for employee
   */
  static __totalTaxAmount(taxable) {
    return taxRate => baseTaxAmount => baseTaxAmount + taxable * taxRate;
  }

  /**
   * Determine pay period (currently monthly) of given date
   * @param {date} payPeriod Date object of current pay period
   * @return {string} Formatted string of pay period from start of
   * month to end of month
   */
  _determinePayPeriod(payPeriod) {
    const payPeriodMoment = moment(payPeriod);
    if (payPeriodMoment.isValid()) {
      const dateFormat = 'DD MMMM';
      const endPayPeriodMoment = payPeriodMoment.clone();
      payPeriodMoment.startOf('month');
      endPayPeriodMoment.endOf('month');

      this._payPeriod = `${payPeriodMoment.format(
        dateFormat
      )} to ${endPayPeriodMoment.format(dateFormat)}`;
    }
  }

  /**
   * Calculate gross based on annual salary / 12 months
   * @return {number} Gross income calculated
   */
  _calculateGross() {
    let grossIncome = 0;

    try {
      grossIncome = this._annualSalary / 12;
    } catch (e) {
      PayslipCalculator.logError(
        '_calculateGross',
        `Salary: ${this._annualSalary}`,
        e
      );
    }

    return grossIncome;
  }

  /**
   * Calculate income tax based on tax limit
   * Method will first reverse the provided TAXABLE_INCOME
   * values (so we can traverse in reverse order), then map over
   * this reversed array to find the correct taxable income based
   * on the employees annual salary
   * @return {number} Income tax based on employee salary and taxable
   * income rules
   */
  _calculateIncomeTax() {
    let totalTax = 0;

    // Avoid mutation/side effect
    const reversedTaxRates = [...TAXABLE_INCOME.default].reverse();

    reversedTaxRates.some(tax => {
      if (this._annualSalary > tax.limit) {
        try {
          const taxable = PayslipCalculator.__totalTaxAmount(
            this._annualSalary - tax.limit
          );
          totalTax = taxable(tax.taxRate)(tax.baseTaxAmount);
        } catch (e) {
          PayslipCalculator.logError(
            'Could not calculate income tax',
            `this._annualSalary: ${this._annualSalary}`,
            `tax object: ${tax}`,
            e
          );
        }
        return true;
      }

      return false;
    });

    return totalTax > 0 ? totalTax / 12 : 0;
  }

  /**
   * Calculates net income based on gross income and income tax
   * @return {number} Net income total
   */
  _calculateNetIncome() {
    return this._grossIncome - this._incomeTax;
  }

  /**
   * Calculates Super based on gross income and super rate
   * @return {number} Super value
   */
  _calculateSuperTotal() {
    const superCalc = this._superRate / 100;
    return this._grossIncome * superCalc;
  }

  /**
   * Public method to generate payslip based on provided variables
   * @return {void}
   * @public
   */
  generate() {
    this._grossIncome = this._calculateGross();
    this._incomeTax = this._calculateIncomeTax();

    this._netIncome = this._calculateNetIncome();
    this._superTotal = this._calculateSuperTotal();
  }
}

export default PayslipCalculator;
