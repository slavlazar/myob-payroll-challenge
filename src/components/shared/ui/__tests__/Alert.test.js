import React from 'react';
import { shallow } from 'enzyme';

import Alert from '../Alert/Alert';
import { Alert as AlertStyle } from '../Alert/Alert.style';
import { renderWithTheme } from '../../../../utils/style-utils';

import theme from '../../../../theme';

const shallowWrapper = (props = {}) => {
  const allProps = {
    children: 'Testing',
    ...props,
  };

  return shallow(<Alert {...allProps} />);
};

describe('Components: Shared: UI: Alert', () => {
  describe('Alert', () => {
    it('renders correctly', () => {
      const wrapped = shallowWrapper();
      expect(wrapped).toMatchSnapshot();
    });

    it('renders a title', () => {
      const wrapped = shallowWrapper({ title: 'Test Title' });
      const header = wrapped.find('.alert-header');
      expect(header).toBeDefined();
      expect(header.text()).toEqual('Test Title');
    });
  });

  describe('AlertStyle', () => {
    it('renders correctly', () => {
      const wrapped = renderWithTheme(<AlertStyle />);
      expect(wrapped).toMatchSnapshot();
    });

    it('should render the correct font size', () => {
      const wrapped = renderWithTheme(<AlertStyle size="small" />);
      expect(wrapped).toHaveStyleRule('font-size', theme.font.size.small);
    });

    it('should render the correct background colour', () => {
      const wrapped = renderWithTheme(<AlertStyle type="success" />);
      expect(wrapped).toHaveStyleRule(
        'background-color',
        theme.colour.alert.success
      );
    });
  });
});
