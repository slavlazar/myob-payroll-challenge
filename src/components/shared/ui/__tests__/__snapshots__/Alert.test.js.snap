// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Shared: UI: Alert Alert renders correctly 1`] = `
<styled.article
  size="default"
  type="default"
>
  <styled.div
    title={null}
    type="default"
  >
    Testing
  </styled.div>
</styled.article>
`;

exports[`Components: Shared: UI: Alert AlertStyle renders correctly 1`] = `
.c0 {
  font-size: 1.6rem;
  background-color: #f5f5f5;
  border-radius: 0.4rem;
  margin-bottom: 2rem;
}

<ThemeProvider
  theme={
    Object {
      "colour": Object {
        "alert": Object {
          "danger": "#fff5f7",
          "default": "#f5f5f5",
          "success": "#00d1b2",
        },
        "background": "#f3f4f5",
        "border": "#efefef",
        "buttonPrimary": "#ffffff",
        "buttonPrimaryBackground": "#00aa65",
        "footerBackground": "#1b2830",
        "heading": "#2e3e4f",
        "navBackground": "#ffffff",
        "primary": "#8241AA",
        "secondary": "#ed005f",
        "text": "#515e6d",
      },
      "font": Object {
        "family": Object {
          "body": "NeutroMYOB-Regular",
          "fallback": "-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif",
          "heading": "NeutroMYOB-Medium",
          "strong": "NeutroMYOB-Demi",
        },
        "size": Object {
          "large": "1.8rem",
          "medium": "1.6rem",
          "small": "1.4rem",
          "xlarge": "2.2rem",
          "xsmall": "1.2rem",
        },
        "variants": Object {
          "body": Object {
            "family": "NeutroMYOB-Regular",
            "src": Array [
              "url(https://assets.digital.myob.com/fonts/NeutroMYOB-Regular.woff2)",
              "url(https://assets.digital.myob.com/fonts/NeutroMYOB-Regular.woff)",
            ],
            "style": "normal",
            "weight": 400,
          },
          "heading": Object {
            "family": "NeutroMYOB-Medium",
            "src": Array [
              "url(https://assets.digital.myob.com/fonts/NeutroMYOB-Medium.woff2)",
              "url(https://assets.digital.myob.com/fonts/NeutroMYOB-Medium.woff)",
            ],
            "style": "normal",
            "weight": 600,
          },
          "strong": Object {
            "family": "NeutroMYOB-Demi",
            "src": Array [
              "url(https://assets.digital.myob.com/fonts/NeutroMYOB-Demi.woff2)",
              "url(https://assets.digital.myob.com/fonts/NeutroMYOB-Demi.woff)",
            ],
            "style": "normal",
            "weight": 700,
          },
        },
        "weight": Object {
          "bold": 600,
          "normal": 400,
        },
      },
      "lineHeight": Object {
        "heading": 1.5,
        "text": 1.7,
      },
      "media": Object {
        "desktop": 768,
        "large": 992,
        "tablet": 576,
      },
    }
  }
>
  <styled.article>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  font-size: ",
              [Function],
              ";
  background-color: ",
              [Function],
              ";
  border-radius: 0.4rem;
  margin-bottom: 2rem;
",
            ],
          },
          "displayName": "styled.article",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "article",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <article
        className="c0"
      />
    </StyledComponent>
  </styled.article>
</ThemeProvider>
`;
